pkg/scheduler/api/devices/nvidia/vgpu/utils.go

```diff
@@ -307,43 +285,89 @@ func checkType(annos map[string]string, d GPUDevice, n ContainerDeviceRequest) b
 	return false
 }
 
+// getGPUDeviceSnapShot is not a strict deep copy, the pointer item is same with origin.
 func getGPUDeviceSnapShot(snap *GPUDevices) *GPUDevices {
 	ret := GPUDevices{
-		Name:   snap.Name,
-		Device: make(map[int]*GPUDevice),
-		Score:  float64(0),
+		Name:    snap.Name,
+		Device:  make(map[int]*GPUDevice),
+		Score:   float64(0),
+		Sharing: snap.Sharing,
 	}
 	for index, val := range snap.Device {
 		if val != nil {
 			ret.Device[index] = &GPUDevice{
-				ID:       val.ID,
-				Node:     val.Node,
-				UUID:     val.UUID,
-				PodMap:   val.PodMap,
-				Memory:   val.Memory,
-				Number:   val.Number,
-				Type:     val.Type,
-				Health:   val.Health,
-				UsedNum:  val.UsedNum,
-				UsedMem:  val.UsedMem,
-				UsedCore: val.UsedCore,
+				ID:          val.ID,
+				Node:        val.Node,
+				UUID:        val.UUID,
+				PodMap:      val.PodMap,
+				Memory:      val.Memory,
+				Number:      val.Number,
+				Type:        val.Type,
+				Health:      val.Health,
+				UsedNum:     val.UsedNum,
+				UsedMem:     val.UsedMem,
+				UsedCore:    val.UsedCore,
+				MigTemplate: val.MigTemplate,
+				MigUsage:    val.MigUsage,
 			}
+			ret.Device[index].MigUsage = deepCopyMigInUse(val.MigUsage)
+			klog.V(4).Infoln("getGPUDeviceSnapShot:", ret.Device[index].UsedMem, val.UsedMem, ret.Device[index].MigUsage, val.MigUsage)
 		}
 	}
 	return &ret
 }
 
+func deepCopyMigInUse(src config.MigInUse) config.MigInUse {
+	dst := config.MigInUse{
+		Index: src.Index,
+	}
+
+	dst.UsageList = make(config.MIGS, len(src.UsageList))
+	for i, usage := range src.UsageList {
+		dst.UsageList[i] = config.MigTemplateUsage{
+			Name:      usage.Name,
+			Memory:    usage.Memory,
+			InUse:     usage.InUse,
+			UsedIndex: make([]int, len(usage.UsedIndex)),
+		}
+		copy(dst.UsageList[i].UsedIndex, usage.UsedIndex)
+	}
+
+	return dst
+}
+
+// getSharingMode by default, we use hami core as the partitioning mode
+func getSharingMode(mode string) string {
+	switch mode {
+	case vGPUControllerMIG:
+		return vGPUControllerMIG
+	default:
+		return vGPUControllerHAMICore
+	}
+}
+
 // checkNodeGPUSharingPredicate checks if a pod with gpu requirement can be scheduled on a node.
 func checkNodeGPUSharingPredicateAndScore(pod *v1.Pod, gssnap *GPUDevices, replicate bool, schedulePolicy string) (bool, []ContainerDevices, float64, error) {
 	// no gpu sharing request
 	score := float64(0)
 	if !checkVGPUResourcesInPod(pod) {
 		return true, []ContainerDevices{}, 0, nil
 	}
+	podSharingMode, ok := pod.Annotations[GPUModeAnnotation]
+	if !ok {
+		podSharingMode = vGPUControllerHAMICore
+	}
+	podSharingMode = getSharingMode(podSharingMode)
+	// if the pod requires mig but the device is not in mig mode, return not fitted;
+	// if the pod does not speficy the sharing mode, then any device mode will be fitted
+	if podSharingMode == vGPUControllerMIG && gssnap.Mode != vGPUControllerMIG {
+		return false, []ContainerDevices{}, 0, nil
```

Comment:
It's better to return a log to show the detailed message.

Reply:
Done